File: 2024/Day2/1.py
def all_increasing_or_decreasing(arr):
  """all increasing or decreasing return true"""
  return all(arr[i] <= arr[i+1] for i in range(len(arr)-1)) or all(arr[i] >= arr[i+1] for i in range(len(arr)-1))

def is_differ_at_least_1_and_at_most_3(arr):
  """is differ at least 1 and at most 3 return true"""
  return min(arr) >= 1 and max(arr) <= 3

with open("input.txt", "r") as file:
  content = file.readlines()
  count = 0
  for line in content:
    arr = list(map(lambda x: int(x), line.strip().split(" ")))
    safe_list = list()
    for i in range(1, len(arr)):
      safe_list.append(abs(arr[i] - arr[i-1]))
    cond1 = all_increasing_or_decreasing(arr)
    cond2 = is_differ_at_least_1_and_at_most_3(safe_list)
    if cond1 and cond2:
      count += 1

  print(count)

File: 2024/Day2/1.swift
import Foundation

func allIncreasingOrDecreasing(_ arr: [Int]) -> Bool {
    // Check if the array is all increasing or all decreasing
    let increasing = zip(arr, arr.dropFirst()).allSatisfy { $0 <= $1 }
    let decreasing = zip(arr, arr.dropFirst()).allSatisfy { $0 >= $1 }
    return increasing || decreasing
}

func isDifferAtLeast1AndAtMost3(_ arr: [Int]) -> Bool {
    // Check if all differences are between 1 and 3 inclusive
    guard let minElement = arr.min(), let maxElement = arr.max() else { return false }
    return minElement >= 1 && maxElement <= 3
}

// Read the input file
let filePath = "input.txt"
var count = 0

do {
    let content = try String(contentsOfFile: filePath, encoding: .utf8)
    let lines = content.split(separator: "\n")

    for line in lines {
        let arr = line.split(separator: " ").compactMap { Int($0) }
        var safeList = [Int]()

        for i in 1..<arr.count {
            safeList.append(abs(arr[i] - arr[i - 1]))
        }

        let cond1 = allIncreasingOrDecreasing(arr)
        let cond2 = isDifferAtLeast1AndAtMost3(safeList)

        if cond1 && cond2 {
            count += 1
        }
    }

    print(count)

} catch {
    print("Error reading file: \(error)")
}
